Name: Shreyas Dongre   
Roll no: 082          
Stream: CS (AI)      
Division: I                         
Semester: 5th Semester                                           
Batch: I-3            
                                                                       Date: 2/9/2022       
Subject: DL
###Experiment-4

In [1]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
df = pd.read_csv('diabetes.csv')

df

Glucose   BMI  Outcome
0        148  33.6        1
1         85  26.6        0
2        183  23.3        1
3         89  28.1        0
4        137  43.1        1
..       ...   ...      ...
763      101  32.9        0
764      122  36.8        0
765      121  26.2        0
766      126  30.1        1
767       93  30.4        0

[768 rows x 3 columns]

In [5]:
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
std_scalar = StandardScaler()
#makes data scale free by conv statistical dist of data
#mean is made 0 and var is made 1
#x=(x-mean)/std
df.iloc[:,[0,1]] = std_scalar.fit_transform(df.iloc[:,[0,1]])
df.head()

Glucose       BMI  Outcome
0  0.848324  0.204013        1
1 -1.123396 -0.684422        0
2  1.943724 -1.103255        1
3 -0.998208 -0.494043        0
4  0.504055  1.409746        1

In [6]:
X=df.iloc[:,0:2].values
Y=df.iloc[:,2].values

In [8]:
x_train,x_test,y_train,y_test = train_test_split(X,Y,test_size=0.2,random_state=np.random.seed(10))

In [11]:
#buliding seqential NN
#i) building 3 layer NN
from keras import layers
from keras import models
model=models.Sequential()
model.add(layers.Dense(16,activation='relu',input_shape=(2,)))
model.add(layers.Dense(16,activation='relu'))
model.add(layers.Dense(1,activation='sigmoid'))


In [12]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 16)                48        
                                                                 
 dense_4 (Dense)             (None, 16)                272       
                                                                 
 dense_5 (Dense)             (None, 1)                 17        
                                                                 
Total params: 337
Trainable params: 337
Non-trainable params: 0
_________________________________________________________________


In [13]:
from tensorflow.keras.optimizers import Adam
model.compile(loss='binary_crossentropy',optimizer=Adam(), metrics=['accuracy'])

In [14]:
model.fit(x_train,y_train,epochs=10)

Epoch 1/10
20/20 [==============================] - 1s 1ms/step - loss: 0.6310 - accuracy: 0.6759
Epoch 2/10
20/20 [==============================] - 0s 1ms/step - loss: 0.5897 - accuracy: 0.7606
Epoch 3/10
20/20 [==============================] - 0s 1ms/step - loss: 0.5573 - accuracy: 0.7606
Epoch 4/10
20/20 [==============================] - 0s 1ms/step - loss: 0.5318 - accuracy: 0.7606
Epoch 5/10
20/20 [==============================] - 0s 1ms/step - loss: 0.5146 - accuracy: 0.7606
Epoch 6/10
20/20 [==============================] - 0s 1ms/step - loss: 0.5032 - accuracy: 0.7590
Epoch 7/10
20/20 [==============================] - 0s 1ms/step - loss: 0.4973 - accuracy: 0.7590
Epoch 8/10
20/20 [==============================] - 0s 1ms/step - loss: 0.4937 - accuracy: 0.7606
Epoch 9/10
20/20 [==============================] - 0s 1ms/step - loss: 0.4923 - accuracy: 0.7622
Epoch 10/10
20/20 [==============================] - 0s 1ms/step - loss: 0.4917 - accuracy: 0.7590


In [16]:
#ii) changing epoch
from keras import layers
from keras import models
model2=models.Sequential()
model2.add(layers.Dense(16,activation='relu',input_shape=(2,)))
model2.add(layers.Dense(16,activation='relu'))
model2.add(layers.Dense(1,activation='sigmoid'))

print(model.summary())

model2.compile(loss='binary_crossentropy',optimizer=Adam(), metrics=['accuracy'])

model2.fit(x_train,y_train,epochs=900)

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 16)                48        
                                                                 
 dense_4 (Dense)             (None, 16)                272       
                                                                 
 dense_5 (Dense)             (None, 1)                 17        
                                                                 
Total params: 337
Trainable params: 337
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/900
20/20 [==============================] - 0s 1ms/step - loss: 0.6489 - accuracy: 0.6612
Epoch 2/900
20/20 [==============================] - 0s 1ms/step - loss: 0.6042 - accuracy: 0.7264
Epoch 3/900
20/20 [==============================] - 0s 1ms/step - loss: 0.5694 - accuracy: 0.7622
Epoch 4/900
20/20 

In [19]:
#iii) changing no of hidden layers
from keras import layers
from keras import models
model3=models.Sequential()
model3.add(layers.Dense(16,activation='relu',input_shape=(2,)))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))
model3.add(layers.Dense(16,activation='relu'))

model3.add(layers.Dense(1,activation='sigmoid'))

print(model.summary())

model3.compile(loss='binary_crossentropy',optimizer=Adam(), metrics=['accuracy'])

model3.fit(x_train,y_train,epochs=200)

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 16)                48        
                                                                 
 dense_4 (Dense)             (None, 16)                272       
                                                                 
 dense_5 (Dense)             (None, 1)                 17        
                                                                 
Total params: 337
Trainable params: 337
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/200
20/20 [==============================] - 1s 1ms/step - loss: 0.6802 - accuracy: 0.6596
Epoch 2/200
20/20 [==============================] - 0s 1ms/step - loss: 0.6398 - accuracy: 0.6596
Epoch 3/200
20/20 [==============================] - 0s 1ms/step - loss: 0.6012 - accuracy: 0.6596
Epoch 4/200
20/20 

In [20]:
#iv) changing no of hidden neurons
from keras import layers
from keras import models
model3=models.Sequential()
model3.add(layers.Dense(16,activation='relu',input_shape=(2,)))
model3.add(layers.Dense(32,activation='relu'))
model3.add(layers.Dense(32,activation='relu'))
model3.add(layers.Dense(32,activation='relu'))
model3.add(layers.Dense(32,activation='relu'))
model3.add(layers.Dense(32,activation='relu'))


model3.add(layers.Dense(1,activation='sigmoid'))

print(model.summary())

model3.compile(loss='binary_crossentropy',optimizer=Adam(), metrics=['accuracy'])

model3.fit(x_train,y_train,epochs=200)

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 16)                48        
                                                                 
 dense_4 (Dense)             (None, 16)                272       
                                                                 
 dense_5 (Dense)             (None, 1)                 17        
                                                                 
Total params: 337
Trainable params: 337
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/200
20/20 [==============================] - 0s 1ms/step - loss: 0.6551 - accuracy: 0.7166
Epoch 2/200
20/20 [==============================] - 0s 1ms/step - loss: 0.5637 - accuracy: 0.7459
Epoch 3/200
20/20 [==============================] - 0s 1ms/step - loss: 0.5156 - accuracy: 0.7573
Epoch 4/200
20/20 

In [23]:
#v) changing activation function
from keras import layers
from keras import models
model3=models.Sequential()
model3.add(layers.Dense(16,activation='sigmoid',input_shape=(2,)))
model3.add(layers.Dense(50,activation='sigmoid'))
model3.add(layers.Dense(50,activation='sigmoid'))
model3.add(layers.Dense(50,activation='sigmoid'))
model3.add(layers.Dense(50,activation='sigmoid'))
model3.add(layers.Dense(50,activation='sigmoid'))
model3.add(layers.Dense(50,activation='sigmoid'))
model3.add(layers.Dense(50,activation='sigmoid'))


model3.add(layers.Dense(1,activation='sigmoid'))

print(model.summary())

model3.compile(loss='binary_crossentropy',optimizer=Adam(), metrics=['accuracy'])

model3.fit(x_train,y_train,epochs=200)

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 16)                48        
                                                                 
 dense_4 (Dense)             (None, 16)                272       
                                                                 
 dense_5 (Dense)             (None, 1)                 17        
                                                                 
Total params: 337
Trainable params: 337
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/200
20/20 [==============================] - 0s 2ms/step - loss: 0.6566 - accuracy: 0.6596
Epoch 2/200
20/20 [==============================] - 0s 1ms/step - loss: 0.6436 - accuracy: 0.6596
Epoch 3/200
20/20 [==============================] - 0s 1ms/step - loss: 0.6429 - accuracy: 0.6596
Epoch 4/200
20/20 

In [ ]:
! jupyter nbconvert --to html I082_Shreyas_Dongre_DL_Exp4.ipynb

#Conclusion:
1) Implemented 3 layer sequential neural network and used binary crossentropy as loss function. Achieved 75.9% accuracy.                         
2) Merely changing epoch and activation function did not help to achieve much accuracy.       
3) Increasing number of hidden neurons and hidden layers(to an extent) helped to achieve 6-7% more accuracy.